dwave/plugins/torch/autoencoder/ae.py

```diff
@@ -0,0 +1,112 @@
+            text-to-image model, x is a sequence of tokens, and x' is an image). Note
+            that the decoder input is of shape (batch_size, n, d1, d2, ...), where n is
+            a number of discrete representations to be created from a single latent
+            representation of a single initial data point.
+        latent_to_discrete (Callable[[torch.Tensor, int], torch.Tensor] | None): A
+            stochastic function that maps the output of the encoder to a discrete
+            representation. Importantly, since the function is stochastic, it allows for
+            the creation of multiple discrete representations from the latent
+            representation of a single data point. Thus, the signature of this function
+            is (l, n) -> d, where l is the output of the encoder and has shape
+            (batch_size, l1, l2, ...), n is the number of discrete representations per
+            data point, and d has shape (batch_size, n, d1, d2, ...), which will be the
+            input to the decoder. If None, the gumbel softmax function is used for
+            stochasticity. Defaults to None.
+    """
+
+    def __init__(
+        self,
+        encoder: torch.nn.Module,
+        decoder: torch.nn.Module,
+        latent_to_discrete: Callable[[torch.Tensor, int], torch.Tensor] | None = None,
+    ):
+        super().__init__()
+        self.encoder = encoder
+        self.decoder = decoder
+        if latent_to_discrete is None:
+
+            def _latent_to_discrete(
+                logits: torch.Tensor, n_samples: int
+            ) -> torch.Tensor:
+                # Logits is of shape (batch_size, n_discrete), we assume these logits
+                # refer to the probability of each discrete variable being 1. To use the
+                # gumbel softmax function we need to reshape the logits to (batch_size,
+                # n_discrete, 1), and then stack the logits to a zeros tensor of the
+                # same shape. This is done to ensure that the gumbel softmax function
+                # works correctly.
+
+                logits = logits.unsqueeze(-1)
+                logits = torch.cat((logits, torch.zeros_like(logits)), dim=-1)
+                # We now create a new leading dimension and repeat the logits n_samples
+                # times:
+                logits = logits.unsqueeze(1).repeat(1, n_samples, 1, 1)
+                one_hots = torch.nn.functional.gumbel_softmax(
+                    logits, tau=1 / 7, hard=True
+                )
+                # one_hots is of shape (batch_size, n_samples, n_discrete, 2), we need
+                # to take the first element of the last dimension and convert it to spin
+                # variables to make the latent space compatible with QPU models.
+                return one_hots[..., 0] * 2 - 1
+
+            latent_to_discrete = _latent_to_discrete
+
+        self.latent_to_discrete = latent_to_discrete
```

Comment:
(for readability) can move this above the conditional and do the `self.` assignment at 102

Reply:
Changed this too.